File: packages/clipboard/ios/Plugin/ClipboardPlugin.swift
import Foundation
import Capacitor

@objc(ClipboardPlugin)
public class ClipboardPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ClipboardPlugin"
    public let jsName = "Clipboard"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "read", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "write", returnType: CAPPluginReturnPromise)
    ]
    public static let tag = "ClipboardPlugin"

    private var implementation: Clipboard?

    override public func load() {
        self.implementation = Clipboard(plugin: self)
    }

    @objc func read(_ call: CAPPluginCall) {
        implementation?.read { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            self.resolveCall(call, result)
        }
    }

    @objc func write(_ call: CAPPluginCall) {
        do {
            let options = try WriteOptions(call)
            implementation?.write(options) { error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call)
            }
        } catch {
            rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", ClipboardPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/clipboard/android/src/main/java/io/capawesome/capacitorjs/plugins/clipboard/ClipboardPlugin.java
package io.capawesome.capacitorjs.plugins.clipboard;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.clipboard.classes.CustomException;
import io.capawesome.capacitorjs.plugins.clipboard.classes.options.WriteOptions;
import io.capawesome.capacitorjs.plugins.clipboard.classes.results.ReadResult;
import io.capawesome.capacitorjs.plugins.clipboard.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.clipboard.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.clipboard.interfaces.Result;

@CapacitorPlugin(name = "Clipboard")
public class ClipboardPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "ClipboardPlugin";

    private Clipboard implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new Clipboard(this);
    }

    @PluginMethod
    public void read(PluginCall call) {
        try {
            NonEmptyResultCallback<ReadResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull ReadResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.read(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void write(PluginCall call) {
        try {
            WriteOptions options = new WriteOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.write(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
